[ES|QL] Support first(), last() functions

mouhc1ne opened the pull request:
**Commit 1**

Create copies of the following classes into their respective, and temporary, `All*` equivalent:
- `FirstLongByTimestampAggregator`
- `LongLongState`
- `First`

**Commit 2**

- Register `ALL_FIRST` as snapshot function, and restrict it to the long type only, for now.
- Exceptionally allow `ALL_FIRST` agg function to be passed down null values by exempting it from the corresponding rule.
- Change agg and grouping agg implementers, to pass all values down to the aggregator which must handle null values according to its internal logic.

**Commit 3**
- Sample csv tests for illustrating purposes only.
- Will be removed and substituted with tests that read from one of the existing datasets instead of creating new one.

mouhc1ne committed:
[ES|QL] Support first(), last() functions

Create copies of the following classes into their respective, and temporary, "All*" equivalent:
- FirstLongByTimestampAggregator
- LongLongState
- First

mouhc1ne committed:
[ES|QL] Support first(), last() functions
- Register ALL_FIRST as snapshot function, and restrict it to the long type only, for now.
- Exceptionally allow ALL_FIRST agg function to be passed down null values by exempting it from the corresponding rule.
- Change agg and grouping agg implementers, to pass all values down to the aggregator which must handle null values according to its internal logic.

mouhc1ne committed:
[ES|QL] Support first(), last() functions
- Sample csv tests for illustrating purposes only.
- Will be removed and substituted with tests that read from one of the existing datasets instead of creating new one.

mouhc1ne committed:
Code review

mouhc1ne committed:
code review

mouhc1ne committed:
Move AllLongLongState out of the generated source package

mouhc1ne committed:
spotless

mouhc1ne committed:
code review

mouhc1ne committed:
CSV tests
 - Insert a null value in the k8s-downsampled dataset in the row with the earlier timestamp
 - Add csv tests that check for the presence of null in both grouping in non-grouping queries
- Revert commit f98b4966 which introduced its own CSV dataset and mapping

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/GroupingAggregatorImplementer.java:
```diff
@@ -468,7 +468,9 @@ private MethodSpec addRawInputLoop(TypeName groupsType, boolean valuesAreVector)
             builder.addStatement("int valuesPosition = groupPosition + positionOffset");
-            if (valuesAreVector == false) {
+
+            // TBD - affects 2 geo classes
```
The effect seems to be to remove a check on null blocks in the generated code. Can I assume this means that the null check is being done higher in the stack?

mouhc1ne replied:
The last change allows the final step of `SpatialExtentGroupingState` to treat infinite values the same way it treats missing ones, by appending null. This allows us to skip a failed attempt at rectangle creation which results in an `IllegalArgException` if no values have been seen for a particular group. Btw, this only happens **after** inserting rows with null shapes in the 2 dataset files, which in itself is only included to exercise the logic inside the 4 `combine` methods and verify if it indeed filters out nulls. At this stage, the csv tests hit 3 out of 4 cases. The missing case is single (no-grouping) for `SpatialExtentGeoShapeDocValuesAggregator` which I couldn't hit yet. Simply removing the group by and adjusting the query doesn't exercise that path for some reason.

mouhc1ne committed:
Code Review
- Place back the skip null/zero values in the Geo aggs, except now this is done in the aggregator instead of the aggregator function.
- Insert missing values in 2 csv datasets to ensure null/zero values continue to be excluded

mouhc1ne committed:
Remove outdated comment.

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -2210,6 +2212,7 @@ count:long | key:keyword | extent:cartesian_shape
 8          | Bot         | BBOX (0.0, 3.0, 1.0, 0.0)
 8          | Top         | BBOX (0.0, 3.0, 3.0, 2.0)
 4          | Fou         | BBOX (0.0, 3.0, 3.0, 0.0)
+1          | Mis         | BBOX (0.0, 0.0, 0.0, 0.0)
```
Also looks wrong.

craigtaverner reviewed (changes requested):
null input should produce null results

mouhc1ne committed:
Changes:
- Allow the final step of SpatialExtentGroupingState to treat infinite values the same way it treats missing ones, by appending null. This allows us to skip a failed attempt at rectangle creation which results in an IllegalArgException if no values have been seen for a particular group

mouhc1ne committed:
code review

mouhc1ne commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -2185,6 +2186,7 @@ count:long | key:keyword | extent:cartesian_shape
 8          | Bot         | BBOX (0.0, 3.0, 1.0, 0.0)
 8          | Top         | BBOX (0.0, 3.0, 3.0, 2.0)
 4          | Fou         | BBOX (0.0, 3.0, 3.0, 0.0)
+1          | Mis         | BBOX (0.0, 0.0, 0.0, 0.0)
```
Returning null now.

nik9000 commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -3024,3 +3027,17 @@ wkt:keyword                |pt:cartesian_point
+# this test ensures that the non-grouping ST_EXTENT_AGG is able to gracefully handle missing cartesian shapes
+
+FROM cartesian_multipolygons
+| STATS extent = ST_EXTENT_AGG(shape), count = COUNT()
```
There are other tests with `ST_EXTENT_AGG(shape)` that'll absorb your new `null` lines without needing a new test. Right?

I suppose you could add a `WHERE shape IS NOT NULL` filter here and name it something else. I imagine this is pretty similar for all of your other tests. Right?

nik9000 reviewed (approved):
LGTM. @

nik9000 reviewed (commented):
Code looks good to me. The tests need another look I think.

mouhc1ne committed:
Changes:
- Filter-out null rows from existing spatial CSV tests that are affected by the new rows containing null shapes.
- Add equivalent CSV tests that operate only the rows (ones with null shapes), and assert that we get nulls.
- Allow SpatialExtentState to return null when a row was seen but X/Y values remain in their default state.

Commit:
[CI] Auto commit changes from spotless

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/InferNonNullAggConstraint.java:
```diff
@@ -56,6 +57,12 @@ protected LogicalPlan rule(Aggregate aggregate, LocalLogicalOptimizerContext con
                 // ignore literals (e.g. COUNT(1))
                 // make sure the field exists at the source and is indexed (not runtime)
+
+                if (af instanceof AllFirst) {
```
:+1:


Not a fan of this, but we'll be less hacky when we have more aggs like this.

nik9000 reviewed (approved):
LGTM.

@craigtaverner should have a look too.

mouhc1ne committed:
Changes:
- Minor adjustments to csv tests.

mouhc1ne committed:
Merge branch 'main' into all-first

mouhc1ne committed:
Revert "Changes:"

This reverts commit 92e3bfb88c5828b0499c80431f64398bb2ea9a6b.

mouhc1ne committed:
Changes:
- Source data from a ROW during ALL_FIRST csv test, without changing existing datasets.

mouhc1ne committed:
Merge branch 'main' into all-first

mouhc1ne committed:
Merge branch 'main' into all-first

mouhc1ne committed:
Merge branch 'main' into all-first

mouhc1ne committed:
Merge branch 'main' into all-first

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java:
```diff
@@ -1652,6 +1652,16 @@ public enum Cap {
+        /**
+         * Allow ST_EXTENT_AGG to gracefully handle missing spatial shapes
+         */
+        ST_EXTENT_AGG_NULL_SUPPORT,
```
Ah, OK, so there were bugs fixed here! Thanks for finding and fixing these!

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -3024,3 +3027,74 @@ wkt:keyword                |pt:cartesian_point
+# this test ensures that the non-grouping ST_EXTENT_AGG is able to gracefully handle missing cartesian shapes
+
+required_capability: st_intersects
+required_capability: st_extent_agg_null_support
```
If there was no bug in ST_INTERSECTS, we don't need this required capability.

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -3024,3 +3027,74 @@ wkt:keyword                |pt:cartesian_point
+1          | Mis         | null
+;
+
+stExtentCartesianShapesNonGroupingOnNulls
```
I noticed all tests are on cartesian data, never geo data. This could be sufficient, since the changes are identical across the two types, but ideally tests on geo would also be nice.

craigtaverner reviewed (approved):
Looks great. Fantastic that you also found a fixed a bug in ST_EXTENT_AGG on null handling! I had some minor comments and questions, but generally like the solution.

mouhc1ne commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/spatial/SpatialExtentCartesianShapeDocValuesAggregator.java:
```diff
@@ -38,10 +38,16 @@ public static SpatialExtentGroupingState initGrouping() {
     }
 
     public static void combine(SpatialExtentState current, @Position int p, IntBlock values) {
+        if (values.getValueCount(p) == 0) {
```
Correct.

mouhc1ne committed:
minor csv test changes

mouhc1ne committed:
Merge branch 'main' into all-first
